lib/Conversion/TorchToLinalg/Linear.cpp

```diff
@@ -1673,6 +1673,581 @@ Value ConvertAtenConvolutionOp::createTransposedInputPadding(
+          continue;
+        }
+        indices.push_back({i});
+      }
+
+      auto retType = inType.clone(makeShapeLLVMCompatible(outShape));
+      return tensor::ExpandShapeOp::create(rewriter, loc, retType, tensor,
+                                           indices);
+    };
+    // The createZeroInitExpandedGroupsTensor lambda function below is used to
+    // create empty tensor with already expanded group dimension.
+    auto createZeroInitExpandedGroupsTensor =
+        [&](OpBuilder &rewriter, Location loc, const SmallVector<Value> &sizes,
+            Type type, int64_t dim,
+            SmallVector<ReassociationIndices> &indices) {
+          Value groups =
+              mlir::arith::ConstantIndexOp::create(rewriter, loc, numGroups);
+
+          SmallVector<Value> expandedSizes;
+          for (auto i = 0; i < static_cast<int64_t>(sizes.size()); i++) {
+            if (i == dim) {
+              expandedSizes.push_back(groups);
+              expandedSizes.push_back(
+                  rewriter.createOrFold<arith::FloorDivSIOp>(loc, sizes[i],
+                                                             groups));
+            } else {
+              expandedSizes.push_back(sizes[i]);
+            }
+          }
+
+          indices.clear();
+          for (auto i = 0; i <= static_cast<int64_t>(sizes.size()); i++) {
+            if (i == dim) {
+              indices.push_back({i, ++i});
+              continue;
+            }
+            indices.push_back({i});
+          }
+
+          return createZeroInitTensor(rewriter, loc, expandedSizes, type);
+        };
+
+    SmallVector<Value> newResults(op->getNumResults());
+
+    // Computing Backward-Input Convolution.
+    if (outMask[0]) {
+      // If convolution bwd is grouped, `grad_output` should be expanded.
+      auto gradOutputExpanded =
+          isGroupedConvBwd ? expandGroups(gradOutput, 1) : gradOutput;
+      // If convolution bwd is grouped, `weight` should be expanded
+      auto weightExpanded = isGroupedConvBwd ? expandGroups(weight, 0) : weight;
+
+      // Flip weight along spatial dims only if
+      // - kernel size is greater than 1,
+      // - the kernel is not a 1x1 or 1x1x1 kernel.
+      SmallVector<int64_t> weightDimsInt = makeShapeTorchCompatible(
+          cast<RankedTensorType>(weightExpanded.getType()).getShape());
+      bool is1x1Kernel = std::all_of(weightDimsInt.rbegin(),
+                                     weightDimsInt.rbegin() + numSpatialDims,
+                                     [](int64_t dim) { return dim == 1; });
+      if (numSpatialDims > 1 && !is1x1Kernel) {
+        SmallVector<int64_t> weightFlipDims;
+        weightFlipDims.reserve(numSpatialDims);
+        for (int64_t i = 0; i < static_cast<int64_t>(numSpatialDims); ++i)
+          weightFlipDims.push_back(spatialStartDimIdx + i);
+        weightExpanded = torch_to_linalg::flipTensor(
+            rewriter, loc, weightExpanded, weightFlipDims);
+      }
```

Comment:
Okay, we also care about 3x1 kernels, and we still won't need to flip the second spatial dim. It may be simpler to do something like:


```suggestion
      SmallVector<int64_t> weightDimsInt = makeShapeTorchCompatible(
          cast<RankedTensorType>(weightExpanded.getType()).getShape());
      # Collect any non-unit spatial dim indices.
      SmallVector<int64_t> weightFlipDims;
      for (auto e in llvm::enumerate(weightDimsInt)) {
         if (e.value() == 1)
           weightFlipDims.push_back(static_cast<int64_t>(e.index());
      }
      # Perform a flip if we have any non-trivial spatial dims.
      if (!weightFlipDims.empty()) {
        weightExpanded = torch_to_linalg::flipTensor(
            rewriter, loc, weightExpanded, weightFlipDims);
      }
```